transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -271,32 +289,70 @@ def _parse_conn_string(self, conn_string):
         self._setup_connection(conn_id, connection_string)
 
         try:
-            self._control_info = await self._jlink_async.verify_control_structure(self._device_info, self._control_info)
+            self._control_info = await self._jlink_async.verify_control_structure(
+                self._device_info, self._control_info)
+
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.CONNECTION_BIT, True)
         except HardwareError as e:
             logger.warning("RPC, streaming and tracing won't work, but device still can be flashed or debugged",
                 exc_info=True)
 
         self._connection_id = conn_id
 
     async def disconnect(self, conn_id):
-        await self._jlink_async.close_jlink()
+        try:
+            for interface in JLinkAdapter.POLLING_INTERFACES:
+                if self.opened_interfaces[interface]:
+                    await self.close_interface(conn_id, interface)
+
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.CONNECTION_BIT, False)
+        except pylink.errors.JLinkReadException:
+            logger.warning("Error disconnecting jlink adapter", exc_info=True)
+
         self._teardown_connection(conn_id)
 
     async def open_interface(self, conn_id, interface):
         self._ensure_connection(conn_id, True)
 
+        if not self._control_info:
```

Comment:
This error chomping should not be necessary.  If there is no control_info and someone tries to open an interface, the attempt should fail because that interface could not be opened.  The bug is with the calling code not this code.  @DimaWittmann Where are you seeing errors that this code is working around because we need to fix those instead.

Reply:
Note as a caveat to this that the `debug` interface does not require control_info (and opening it is a no-op).  For all other interfaces, opening/closing them should fail if there is no config info

Reply:
I see your point, I will review these checks